File: sbe/binance/spot_3_2/generated-php/spot_sbe/AccountCommissionResponse.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class AccountCommissionResponse
{
    public const TEMPLATE_ID = 405;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 2;
    public const BLOCK_LENGTH = 108;

    public int|float|array|null $commissionExponent = null;
    public int|float|array|null $discountExponent = null;
    public int|float|array|null $standardCommissionMaker = null;
    public int|float|array|null $standardCommissionTaker = null;
    public int|float|array|null $standardCommissionBuyer = null;
    public int|float|array|null $standardCommissionSeller = null;
    public int|float|array|null $taxCommissionMaker = null;
    public int|float|array|null $taxCommissionTaker = null;
    public int|float|array|null $taxCommissionBuyer = null;
    public int|float|array|null $taxCommissionSeller = null;
    public int|float|array|null $discountEnabledForAccount = null;
    public int|float|array|null $discountEnabledForSymbol = null;
    public int|float|array|null $discount = null;
    public int|float|array|null $specialCommissionMaker = null;
    public int|float|array|null $specialCommissionTaker = null;
    public int|float|array|null $specialCommissionBuyer = null;
    public int|float|array|null $specialCommissionSeller = null;
    public string $symbol = '';
    public string $discountAsset = '';

    public function encode(): string
    {
        $buffer = '';

        if ($this->commissionExponent !== null) {
            $buffer .= pack('c', $this->commissionExponent);
        }
        if ($this->discountExponent !== null) {
            $buffer .= pack('c', $this->discountExponent);
        }
        if ($this->standardCommissionMaker !== null) {
            $buffer .= pack('q', $this->standardCommissionMaker);
        }
        if ($this->standardCommissionTaker !== null) {
            $buffer .= pack('q', $this->standardCommissionTaker);
        }
        if ($this->standardCommissionBuyer !== null) {
            $buffer .= pack('q', $this->standardCommissionBuyer);
        }
        if ($this->standardCommissionSeller !== null) {
            $buffer .= pack('q', $this->standardCommissionSeller);
        }
        if ($this->taxCommissionMaker !== null) {
            $buffer .= pack('q', $this->taxCommissionMaker);
        }
        if ($this->taxCommissionTaker !== null) {
            $buffer .= pack('q', $this->taxCommissionTaker);
        }
        if ($this->taxCommissionBuyer !== null) {
            $buffer .= pack('q', $this->taxCommissionBuyer);
        }
        if ($this->taxCommissionSeller !== null) {
            $buffer .= pack('q', $this->taxCommissionSeller);
        }
        if ($this->discount !== null) {
            $buffer .= pack('q', $this->discount);
        }
        if ($this->specialCommissionMaker !== null) {
            $buffer .= pack('q', $this->specialCommissionMaker);
        }
        if ($this->specialCommissionTaker !== null) {
            $buffer .= pack('q', $this->specialCommissionTaker);
        }
        if ($this->specialCommissionBuyer !== null) {
            $buffer .= pack('q', $this->specialCommissionBuyer);
        }
        if ($this->specialCommissionSeller !== null) {
            $buffer .= pack('q', $this->specialCommissionSeller);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->commissionExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->discountExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->standardCommissionMaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->standardCommissionTaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->standardCommissionBuyer = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->standardCommissionSeller = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->taxCommissionMaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->taxCommissionTaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->taxCommissionBuyer = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->taxCommissionSeller = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->discount = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->specialCommissionMaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->specialCommissionTaker = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->specialCommissionBuyer = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->specialCommissionSeller = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
    }
}

File: sbe/binance/spot_3_1/generated-python/spot_sbe/AccountCommissionResponse.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class AccountCommissionResponse:
    """SBE message: AccountCommissionResponse."""

    TEMPLATE_ID = 405
    SCHEMA_ID = 3
    SCHEMA_VERSION = 1
    BLOCK_LENGTH = 108

    def __init__(self):
        self.commission_exponent = None
        self.discount_exponent = None
        self.standard_commission_maker = None
        self.standard_commission_taker = None
        self.standard_commission_buyer = None
        self.standard_commission_seller = None
        self.tax_commission_maker = None
        self.tax_commission_taker = None
        self.tax_commission_buyer = None
        self.tax_commission_seller = None
        self.discount_enabled_for_account = None
        self.discount_enabled_for_symbol = None
        self.discount = None
        self.special_commission_maker = None
        self.special_commission_taker = None
        self.special_commission_buyer = None
        self.special_commission_seller = None
        self.symbol = b''
        self.discount_asset = b''

    def _decode_var_data(self, data: bytes, offset: int) -> Tuple[bytes, int]:
        """Decode variable-length binary data."""
        pos = offset
        length = struct.unpack_from('<I', data, pos)[0]
        pos += 4
        value = data[pos:pos+length]
        pos += length
        return (value, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.commission_exponent is not None:
            buffer.write(struct.pack('<b', self.commission_exponent))
        if self.discount_exponent is not None:
            buffer.write(struct.pack('<b', self.discount_exponent))
        if self.standard_commission_maker is not None:
            buffer.write(struct.pack('<q', self.standard_commission_maker))
        if self.standard_commission_taker is not None:
            buffer.write(struct.pack('<q', self.standard_commission_taker))
        if self.standard_commission_buyer is not None:
            buffer.write(struct.pack('<q', self.standard_commission_buyer))
        if self.standard_commission_seller is not None:
            buffer.write(struct.pack('<q', self.standard_commission_seller))
        if self.tax_commission_maker is not None:
            buffer.write(struct.pack('<q', self.tax_commission_maker))
        if self.tax_commission_taker is not None:
            buffer.write(struct.pack('<q', self.tax_commission_taker))
        if self.tax_commission_buyer is not None:
            buffer.write(struct.pack('<q', self.tax_commission_buyer))
        if self.tax_commission_seller is not None:
            buffer.write(struct.pack('<q', self.tax_commission_seller))
        if self.discount is not None:
            buffer.write(struct.pack('<q', self.discount))
        if self.special_commission_maker is not None:
            buffer.write(struct.pack('<q', self.special_commission_maker))
        if self.special_commission_taker is not None:
            buffer.write(struct.pack('<q', self.special_commission_taker))
        if self.special_commission_buyer is not None:
            buffer.write(struct.pack('<q', self.special_commission_buyer))
        if self.special_commission_seller is not None:
            buffer.write(struct.pack('<q', self.special_commission_seller))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.commission_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.discount_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.standard_commission_maker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.standard_commission_taker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.standard_commission_buyer = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.standard_commission_seller = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.tax_commission_maker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.tax_commission_taker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.tax_commission_buyer = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.tax_commission_seller = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.discount = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.special_commission_maker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.special_commission_taker = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.special_commission_buyer = struct.unpack_from('<q', data, pos)[0]
        pos += 8
        self.special_commission_seller = struct.unpack_from('<q', data, pos)[0]
        pos += 8

        # Skip to end of block for forward compatibility
        pos = 108


        self.symbol, pos = self._decode_var_data(data, pos)
        self.discount_asset, pos = self._decode_var_data(data, pos)
